lib/connection_tester.rb

```diff
@@ -0,0 +1,277 @@
+        ct.nodeinfo(res)
+
+      rescue Failure => e
+        result_from_failure(res, e)
+      end
+
+      res.freeze
+    end
+
+    # assemble a hash of options for a Net::HTTP request.
+    # This also includes a :host key containing the hostname of the server
+    #
+    # @param [String] target URL
+    def request_opts_for_url(url)
+      uri = URI.parse(url)
+      raise AddressFailure,
+            "invalid protocol: '#{uri.scheme.upcase}'" unless a_http?(uri)
+      ssl = (uri.scheme == "https") ? true : false
+      port = ssl ? 443 : 80
+
+      REQUEST_OPTS.merge(
+        host:    uri.host,
+        use_ssl: ssl,
+        port:    port,
+        path:    uri.path
+      )
+    rescue URI::InvalidURIError => e
+      raise AddressFailure, e.message
+    end
+
+    protected
+
+    def result_from_failure(result, error)
+      result.error = error
+
+      case error
+      when AddressFailure, DNSFailure, NetFailure
+        result.reachable = false
+      when SSLFailure
+        result.reachable = true
+        result.ssl_status = false
+      when HTTPFailure
+        result.reachable = true
+      when NodeInfoFailure
+        result.diaspora_version = ""
+      end
+    end
+
+    def a_http?(uri)
+      uri.is_a?(URI::HTTP) || uri.is_a?(URI::HTTPS)
+    end
+  end
+
+  def initialize(hostname)
+    @hostname = hostname
+    @max_redirects = 3
+  end
+
+  # Perform the DNS query
+  #
+  # @param [Result]
+  # @return [Resolv::IPv4, Resolv::IPv6] the resolved IP address
+  def resolve(result)
+    res = Resolv::DNS.new
+    addr = res.getaddress(@hostname)
+    result.ip = addr.to_s
+  rescue Resolv::ResolvError, Resolv::ResolvTimeout => e
+    raise DNSFailure, "'#{@hostname}' - #{e.message}"
+  ensure
+    res.close
+  end
+
+  # Perform a HTTP GET request to determine the following information
+  #  * is the host reachable
+  #  * is port 80/443 open
+  #  * is the SSL certificate valid (only on HTTPS)
+  #  * does the server return a successful HTTP status code
+  # (can't do a HEAD request, since that's not a defined route in the app)
+  # follow up to 3 redirects by default
+  #
+  # @param [Hash] request options
+  # @param [Result]
+  # @return [Integer] HTTP status code
+  def request(opts, result, host=nil, redirect=false)
+    host ||= @hostname
+    @max_redirects -= 1 if redirect
+    raise HTTPFailure, "too many redirects on '#{host}'" if @max_redirects <= 0
+
+    port = opts.delete(:port)
+    path = opts.delete(:path)
+    @uses_ssl = opts[:use_ssl]
+    Net::HTTP.start(host, port, nil, nil, nil, opts) do |http|
+      resp = capture_response_time(http, result, path)
+      handle_http_response(resp, result)
+    end
+  rescue Timeout::Error, Errno::ENETUNREACH, Errno::EHOSTUNREACH, Errno::ECONNREFUSED => e
+    raise NetFailure, e.message
+  rescue OpenSSL::SSL::SSLError => e
+    raise SSLFailure, e.message
+  rescue Net::ProtocolError, ArgumentError
+    raise HTTPFailure, e.message
+  end
```

Comment:
We use  typhoeus for posting mesages and faraday for the Webfinger fetch, using net/http here feels wrong, I'd much prefer faraday.